zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -443,14 +483,66 @@ pub(crate) fn insert_queue_entries<'a>(
     Ok(())
 }
 
-pub(crate) fn replace_queue_entries(
+/// A trait that abstracts over the construction of wallet errors.
+///
+/// In order to make it possible to use [`replace_queue_entries`] in database migrations as well as
+/// in code that returns `SqliteClientError`, it is necessary for that method to be polymorphic in
+/// the error type.
+pub(crate) trait WalletError {
+    fn db_error(err: rusqlite::Error) -> Self;
+    fn corrupt(message: String) -> Self;
+    fn chain_height_unknown() -> Self;
+    fn commitment_tree(err: ShardTreeError<commitment_tree::Error>) -> Self;
+}
+
+impl WalletError for SqliteClientError {
+    fn db_error(err: rusqlite::Error) -> Self {
+        SqliteClientError::DbError(err)
+    }
+
+    fn corrupt(message: String) -> Self {
+        SqliteClientError::CorruptedData(message)
+    }
+
+    fn chain_height_unknown() -> Self {
+        SqliteClientError::ChainHeightUnknown
+    }
+
+    fn commitment_tree(err: ShardTreeError<commitment_tree::Error>) -> Self {
+        SqliteClientError::CommitmentTree(err)
+    }
+}
+
+impl WalletError for WalletMigrationError {
+    fn db_error(err: rusqlite::Error) -> Self {
+        WalletMigrationError::DbError(err)
+    }
+
+    fn corrupt(message: String) -> Self {
+        WalletMigrationError::CorruptedData(message)
+    }
+
+    fn chain_height_unknown() -> Self {
+        WalletMigrationError::CorruptedData(
+            "Wallet migration requires a valid account birthday.".to_owned(),
+        )
+    }
+
+    fn commitment_tree(err: ShardTreeError<commitment_tree::Error>) -> Self {
+        WalletMigrationError::CommitmentTree(err)
+    }
+}
+
+pub(crate) fn replace_queue_entries<E: WalletError>(
     conn: &rusqlite::Transaction<'_>,
     query_range: &Range<BlockHeight>,
     entries: impl Iterator<Item = ScanRange>,
-) -> Result<(), SqliteClientError> {
+    force_rescans: bool,
+) -> Result<(), E> {
     let (to_create, to_delete_ends) = {
-        let mut suggested_stmt = conn.prepare_cached(
-            "SELECT block_range_start, block_range_end, priority
+        let mut suggested_stmt = conn
+            .prepare_cached(
+                "SELECT block_range_start, block_range_end, priority
             FROM scan_queue
```

Comment:
rustfmt has messed up the indentation of the SQL statement.

Reply:
Fixed in https://github.com/zcash/librustzcash/pull/935